extensions/controllers/sandboxclaim_controller.go

```diff
@@ -76,26 +81,70 @@ func (r *SandboxClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request
 		return ctrl.Result{}, nil
 	}
 
-	// cache the original status from sandboxclaim
+	// Cache original status
 	originalClaimStatus := claim.Status.DeepCopy()
+
 	var err error
 	var sandbox *v1alpha1.Sandbox
 	var template *extensionsv1alpha1.SandboxTemplate
 
-	// Try getting template
-	if template, err = r.getTemplate(ctx, claim); err == nil || k8errors.IsNotFound(err) {
-		// This ensures the firewall is up before the pod starts.
-		if npErr := r.reconcileNetworkPolicy(ctx, claim, template); npErr != nil {
-			return ctrl.Result{}, fmt.Errorf("failed to reconcile network policy: %w", npErr)
+	// Check Expiration
+	// The Claim is the source of truth for its own expiration logic.
+	claimExpired := false
+	if claim.Spec.Lifecycle != nil && claim.Spec.Lifecycle.ShutdownTime != nil {
+		claimExpired = time.Now().After(claim.Spec.Lifecycle.ShutdownTime.Time)
+	}
+
+	// 1. Manage Sandbox Resources
+	if !claimExpired {
+		// Logic for active Claims
+		if template, err = r.getTemplate(ctx, claim); err == nil || k8errors.IsNotFound(err) {
+			// Ensure NetworkPolicy
+			if npErr := r.reconcileNetworkPolicy(ctx, claim, template); npErr != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to reconcile network policy: %w", npErr)
+			}
+			sandbox, err = r.getOrCreateSandbox(ctx, claim, template)
 		}
+	} else {
+		// Logic for expired Claims
+		// We only check if the Sandbox exists to update status.
+		existingSandbox := &v1alpha1.Sandbox{}
+		if err = r.Get(ctx, req.NamespacedName, existingSandbox); err == nil {
+			sandbox = existingSandbox
+		} else if k8errors.IsNotFound(err) {
+			// It is expected that the sandbox might be gone if expired
+			err = nil
+			sandbox = nil
+		}
+	}
 
-		// Try getting sandbox even if template is not found
-		// It is possible that the template was deleted after the sandbox was created
-		sandbox, err = r.getOrCreateSandbox(ctx, claim, template)
+	// 2. Logic for Auto-Deletion (Policy=Delete)
+	// We trust the Claim's expiration state. If the Claim is expired and Policy is Delete,
+	// we delete the Claim. We do not need to wait for the Sandbox to reflect this state,
+	// nor do we need to check if the Sandbox is missing.
+	// This avoids race conditions where we might re-create a Sandbox because we haven't deleted the Claim yet.
+	if err == nil && claimExpired && claim.Spec.Lifecycle != nil && claim.Spec.Lifecycle.ShutdownPolicy == extensionsv1alpha1.ShutdownPolicyDelete {
```

Comment:
Would suggest moving this under else branch L108

Reply:
done